### ISSUE TITLE
multiple primary key error when using composite PK with createIfNotExists

### ISSUE THREAD

**Issue by remijj** [slick/slick#1999]
Tips for submitting an issue with slick (read before you submit an issue):

Note: Delete these instructions from the description field before you submit the issue.

1. **Do not use github issues for support questions.** For those, first study the documentation and if they do not contain an answer, use one of the community forums (see http://slick.lightbend.com/community/). Support questions posted as issues will be deleted without reply or further explanation.

2. If you have a feature request, note that slick is a community project. As such, you need to get somebody to implement your idea and make a pull request.

3. If you have a bug report, use the applicable parts of the following template (modify as needed):

**DESCRIPTION:**

Give a short, clear description of the issue. Make sure the title also describes the issue as well as possible.

**STEPS TO REPRODUCE:**

Describe steps on how to reproduce the bug. Include a bare-bones test case (as small as possible) that demonstrates the bug.

**EXPECTED RESULT:**

Describe what you expected to see and why.

**ACTUAL RESULT:**

Describe what you saw. If you got compiler errors or exceptions, include them. When including an exception stack traces, make sure to include also the root cause (search for "Caused by:" in the stack trace).

**SETUP:**

Describe your set-up (database used, profile, configuration, etc)

As with feature requests, note that slick is a community project and the best way to get bugs fixed is to submit a pull request for fixing it, along with a test case that demonstrates the bug existed prior to the PR and does not after the PR.

**Comment by enpuliu** [comment]
The SQL command from 
idsAndNames.schema.createIfNotExistsStatements.mkString
is 
create table if not exists `mickey`.`mytable` (`id` VARCHAR(36) NOT NULL,`name` VARCHAR(20) NOT NULL,`desc` VARCHAR(100) NOT NULL)alter table `mickey`.`mytable` add constraint `pk_a` primary key(`id`,`name`)

There are two SQL commands generated: **CREATE TABLE** and **ALTER TABLE**, and there is a "**;**" missing between them. so a SQL syntax error occurred.

**Comment by szeiger** [comment]
The problem is here: https://github.com/slick/slick/blob/b0a43942ae2e7a2ebb6fbf048da20fb3e0a1178d/slick/src/main/scala/slick/jdbc/JdbcStatementBuilderComponent.scala#L608

It passes a `checkNotExists` flag to `createTable` but then calls `createPrimaryKey` and `createIndex` which do not even have such a flag.

It's not limited to compound primary keys, either. Any explicit PK or index will do the job. (non-compound PKs are usually defined by setting a flag in the column instead of creating an explicit `Index` object)

**Comment by ASchmidt84** [comment]
Okay also here, is their any solution in sight?

**Comment by hvesalai** [comment]
I don't believe anybody is working on it. If you want to take a stab at it, a PR is always welcomed.

**Comment by NicolasRouquette** [comment]
Here is the workaround that I use in my Lagom 1.6 application where I experienced this bug.

Before Slick had support for `.createIfNotExists`, the Lagom 1.4 doc described a workaround:
https://www.lagomframework.com/documentation/1.4.x/scala/ReadSideSlick.html#Global-prepare

Unfortunately, the Lagom 1.4 workaround no longer works with Slick 3.3.2.
I tried a few variants of the metadata query below:
- `MTable.getTables`
- `MTables.getTables("offset")`
- `MTables.getTables("%")`

All of them return nothing so the schemas are always created which becomes an error if they already exist.
My workaround involves ignoring these errors with a `.recoverWith`. It's ugly but it works.

```scala
import slick.jdbc.MySQLProfile.api._
import slick.jdbc.meta.MTable

    database
      .run(
        MTable
          .getTables("%")
          .flatMap {
            tables =>
              log.debug(s"Found ${tables.size} tables named '%'.")
              if (!tables.exists(_.name.name == offsetTable.baseTableRow.tableName)) {
                log.debug(s"Creating schemas...")
                Array(offsetTable.schema,
                      workflowsTable.schema,
                      workflowSessionsTable.schema,
                      reportsTable.schema,
                      groupsTable.schema,
                      publishedDocumentsTable.schema)
                  .reduceLeft(_ ++ _)
                  .create
              } else {
                log.debug(s"Schemas already created.")
                DBIO.successful(())
              }
          }
          .transactionally
      )
      .map(_ => Done)
      .recoverWith {
        t: Throwable =>
          log.error("Assuming that the schemas exist since creation failed.", t)
          Future.successful(Done)
      }
```

If one doesn't care to check w/ the `MTable` metadata query, one could simplify the above to something like this:

```scala
import slick.jdbc.MySQLProfile.api._
import slick.jdbc.meta.MTable

    database
      .run({
        log.debug(s"Creating schemas...")
        Array(offsetTable.schema,
              workflowsTable.schema,
              workflowSessionsTable.schema,
              reportsTable.schema,
              groupsTable.schema,
              publishedDocumentsTable.schema)
          .reduceLeft(_ ++ _)
          .create

      }.transactionally)
      .map(_ => Done)
      .recoverWith {
        t: Throwable =>
          log.error("Assuming that the schemas exist since creation failed.", t)
          Future.successful(Done)
      }
```

- Nicolas.
